### PR TITLE
Extension of smartbanner.js to support OS and device type related APP links

### PR DESCRIPTION
**Pull request by Ricco1**
#### Device type related metatags

In one of my projects I came across the situation when app developers have created separate Apps for Phone and Tablet devices, both for iOS and Android. In the current smartbanner.js revision it is only possible to add one **Global** link, which was not suitable for my situation. So I did an extension to the current smartbanner.js revision to allow also to set the separate link for the iOS device combinations. So, as a result the current App links related metas should look like this:

```html
<meta name="smartbanner:button-url-apple" content="https://itunes.apple.com/us/genre/ios/id36?mt=8">
<meta name="smartbanner:button-url-apple-iphone" content="https://itunes.apple.com/iphonelink">
<meta name="smartbanner:button-url-apple-ipad" content="https://itunes.apple.com/ipadlink">
<meta name="smartbanner:button-url-google" content="https://play.google.com/store">
<meta name="smartbanner:button-url-google-phone" content="https://play.google.com/phonelink">
<meta name="smartbanner:button-url-google-tablet" content="https://play.google.com/tabletlink">
```

So now, if you have only one link available then it is still possible to use the old approach of setting **Global** link, while if there are several links for one OS intended for different devices you can add those by omitting the Global link meta tag. If for some reason user will leave global link and add some links to devices specific meta tags, then the latter will be ignored and the **Global** will still be used by the app.

### CODE REVIEW

**Comment by Ricco1** [comment]
Should be OK now. The only issue I am still facing is to run **npm run bannerise**, where I get error as on windows `add-banner --template=banner.tmpl dist/*.*s` was failing for me constantly, so as a result your banners will not be in my revision.

**Comment by Ricco1** [comment]
Hi Ain,

Did you had a chance to look on this?

KR,

Ricco

File: src/smartbanner.js [hunk 1]
```diff
@@ -74,6 +74,7 @@ export default class SmartBanner {
     let parser = new OptionParser();
     this.options = parser.parse();
     this.platform = Detector.platform();
+    this.device = Detector.device();
```
Comment (ain) [inline src/smartbanner.js]: Good call, thanks.

File: src/smartbanner.js [hunk 2]
```diff
@@ -106,11 +107,31 @@ export default class SmartBanner {
   }
 
   get buttonUrl() {
-    if (this.platform === 'android') {
-      return this.options.buttonUrlGoogle;
-    } else if (this.platform === 'ios') {
-      return this.options.buttonUrlApple;
+    let o = this.options;
```
Comment (ain) [inline src/smartbanner.js]: Let's not call things `o` or `ä`. We don't want to document code inline so let's keep to fully readable variable names regardless of scope. Could we use `options`?

File: test/spec/detector_spec.js [hunk 3]
```diff
@@ -27,19 +28,25 @@ describe('Detector', function() {
 
   describe('platform', function() {
 
-    let platform = null;
+    let platform = null,
+        device = null;
```
Comment (ain) [inline test/spec/detector_spec.js]: Let's keep standard variable definitions as per line. I know why you're using it this way, but let's stick to original convention so coding standard doesn't vary as much across codebase. One line, one definition.

**Comment by ain** [comment]
@Ricco1 Re bannerising, please report another bug and let's see what we can do about it for the people on Windows.

**Comment by Ricco1** [comment]
Hi ain,

Changes done as requested and bug reported.

KR,

Ricco

File: README.md [hunk 4]
```diff
@@ -33,7 +34,11 @@ Customisable smart app banner for iOS and Android.
 <meta name="smartbanner:icon-google" content="https://url/to/google-play-icon.png">
 <meta name="smartbanner:button" content="VIEW">
 <meta name="smartbanner:button-url-apple" content="https://ios/application-url">
+<meta name="smartbanner:button-url-apple-iphone" content="https://ios-iphone/application-url">
```
Comment (ain) [inline README.md]: Let's move all these specific cases out of _Basic Usage_ to _Advanced Usage_. We'd like to keep the basic block as simple as possible for easy entry.

File: README.md [hunk 5]
```diff
@@ -60,6 +65,22 @@ In this case you can define a regular expression, which matches all user agent s
...
+```html
+  <meta name="smartbanner:button-url-apple-iphone" content="https://itunes.apple.com/iphonelink">
+```
+
+#### iOS - iPad app link can be set with the following metatag:
+```html
+    <meta name="smartbanner:button-url-apple-ipad" content="https://itunes.apple.com/ipadlink">
+```
+
+**Note:** Is is important that if 'global' links, like `<meta name="smartbanner:button-url-apple"/>` or `<meta name="smartbanner:button-url-google"/>` will be set, device related app links will be omitted (-phone; -tablet).
```
Comment (ain) [inline README.md]: I did test your code on iPhone and iPad and I did not succeed. Link turned out to be `http://127.0.0.1:8080/undefined` when running the local server (`npm run server`) of your PR code.

Also, considering how we're dealing with hiding the banner on some User Agents, I'd call for not requiring the omission of global link, but using it as fallback if none of the more specific definitions are satisfactory, e.g. on iPod.

File: README.md [hunk 6]
```diff
@@ -60,6 +65,22 @@ In this case you can define a regular expression, which matches all user agent s
 ```
 This regular expression would match any user agent string, that ends with *My Example Webapp*.
 
+### Add a separate links specific to some device/platform
```
Comment (ain) [inline README.md]: Please move all these new sections to the end of _Advanced Usage_ to not to break the connected context of other advanced topics.

File: README.md (hunk 5 above)
Comment (Ricco1) [inline README.md]: Strange, I tested it on my Android and it works fine - while the logic for iPhone is quite similar, except the userAgent of course. Unfortunately I don't have an iPhone nor iPad, so I don't have a chance to test it against real devices, but in Chrome Developer tools against different devices, including iPhone and iPad it works fine as well.

Also the tests are passing against specific userAgent, so not sure why in your case it was not working.

Anyway, I will take a look on option to include a Global link as fallback.

**Comment by Ricco1** [comment]
Hi @ain ,

Hopefully now it should be OK. Still tested again with Chrome userAgent emulation and it worked fine, so hopefully it will be fine for you now as well, besides the tests are passing against these specific userAgents.

P.S.

Logic amended according your suggestion to use global links as fallback.

File: src/smartbanner.js [hunk 7]
```diff
@@ -157,7 +190,17 @@ export default class SmartBanner {
     return Detector.userAgentMatchesRegex(this.options.includeUserAgentRegex);
   }
 
-  publish() {
+  publish(ignoreMetas) {
```
Comment (ain) [inline src/smartbanner.js]: Publish method as the main method should remain clean without any parameters. It's important to keep API clean. `Detector` class should be responsible for any detection, this method here just deals with publishing, not detection.

File: src/smartbanner.js [hunk 8]
```diff
@@ -106,11 +107,43 @@ export default class SmartBanner {
...
+    if (device === 'ipad') {
+      return options.buttonUrlAppleIpad ? options.buttonUrlAppleIpad : options.buttonUrlApple;
+    } else if (device === 'iphone' || device === 'ipod') {
+      return options.buttonUrlAppleIphone ? options.buttonUrlAppleIphone : options.buttonUrlApple;
+    } else if (device === 'phone') {
+      return options.buttonUrlGooglePhone ? options.buttonUrlGooglePhone : options.buttonUrlGoogle;
+    } else if (device === 'tablet') {
+      return options.buttonUrlGoogleTablet ? options.buttonUrlGoogleTablet : options.buttonUrlGoogle;
+    }
+
```
Comment (ain) [inline src/smartbanner.js]: This logic should be simplified, it's too long. `device` and `platform` field combination from `Detector` should become sufficient to decide what to return.

File: src/smartbanner.js (hunk 7 above)
Comment (Ricco1) [inline src/smartbanner.js]: Ok @ain , I will take a look on this.

File: src/smartbanner.js (hunk 8 above)
Comment (Ricco1) [inline src/smartbanner.js]: I tested this today in the office on our iPad mini 4 and it worked perfectly fine for me. Also my implementation was used in the project, which I am working on and our QA team also verified that all 4 cases are working correctly, so I am quite surprised that it is not working for you.
Can you do a check of your userAgent and return the result? [http://www.whoishostingthis.com/tools/user-agent/](url)

**Comment by Ricco1** [comment]
BTW, I think I forgot to do the `npm run build` before deployment, which is the reason why it doesn't work for you :). Can you try it on your side or I can do the build and then do a push.

**Comment by Ricco1** [comment]
Hi @ain,

Amended the logic for ignoring device related meta tags, also did `npm run build` so now you shouldn't have any issues with iPad. As I mentioned above, my implementation was also tested by our QA team and it was working fine, so the problem before was just a missing `npm run build` before push.

**Comment by Ricco1** [comment]
Hi @ain ,
Did you have a chance to take a look on the latest changes?

Also I've added another new functionality that was useful in my case - to allow for an app to call `smartbanner.publilsh()` when necessary. In my case it is showing this banner only for registered users after they log in to system.

KR,
Ricco

**Comment by Ricco1** [comment]
Its not about the other pages, than the others - as you know in SPA there is not one page or another, header part for all pages is the same. Besides, you can use the same page as registered or anonymous user, so it is more about the user privileges than pages and there is no way how to manages this with current solution.

In any case I agree with you that this is might not be in the scope of this pull request, although I couldn't make another until you approve this one, so I will cancel last changes.

**Comment by ain** [comment]
@Ricco1 there's already #42 that covers the #16 and your concern. I'd rather see that work picked up and finalised.

**Comment by Ricco1** [comment]
Ok, I rolled it back to the previous Commit, so should be fine now.
What regards your proposed unfinished pull requests, I will take a look.

KR,

Ricco

**Comment by Ricco1** [comment]
Hi @ain ,

I am closing this pull request as I have to make other changes to my repo on which my project is depending on, but I can't wait anymore for you to at least approve these, so that there is no "out of scope changes".

**Comment by ain** [comment]
The merges have to happen granularly, but if you of course decide to focus your work around your fork, that is fine.

**Pull request closed without merging**